From fc9ab7c7193968839ace9141b63fd7cb728454aa Mon Sep 17 00:00:00 2001
From: html6405 <peter.schelchshorn@mhs-solutions.at>
Date: Sat, 3 Aug 2024 21:34:57 +0200
Subject: [PATCH 07/17] Revert "vulkan_renderengine: Check for support first"

This reverts commit f3369eda0af91df158d66f1e4d906a0f0b1496e7.

(cherry picked from commit 066d06063ab1b4f043fd8f0e51cd5060a32da5e3)
---
 .../include/renderengine/RenderEngine.h       |  2 --
 libs/renderengine/skia/SkiaVkRenderEngine.cpp | 27 ++++++-------------
 libs/renderengine/skia/SkiaVkRenderEngine.h   |  2 ++
 libs/renderengine/tests/RenderEngineTest.cpp  |  8 +++++-
 services/surfaceflinger/SurfaceFlinger.cpp    |  7 +++--
 5 files changed, 20 insertions(+), 26 deletions(-)

diff --git a/libs/renderengine/include/renderengine/RenderEngine.h b/libs/renderengine/include/renderengine/RenderEngine.h
index de05268a67..7047358e62 100644
--- a/libs/renderengine/include/renderengine/RenderEngine.h
+++ b/libs/renderengine/include/renderengine/RenderEngine.h
@@ -104,8 +104,6 @@ public:
 
     static std::unique_ptr<RenderEngine> create(const RenderEngineCreationArgs& args);
 
-    static bool canSupport(GraphicsApi);
-
     virtual ~RenderEngine() = 0;
 
     // ----- BEGIN DEPRECATED INTERFACE -----
diff --git a/libs/renderengine/skia/SkiaVkRenderEngine.cpp b/libs/renderengine/skia/SkiaVkRenderEngine.cpp
index eb7a9d5bfa..40ff364e3e 100644
--- a/libs/renderengine/skia/SkiaVkRenderEngine.cpp
+++ b/libs/renderengine/skia/SkiaVkRenderEngine.cpp
@@ -287,7 +287,6 @@ static GrVkGetProc sGetProc = [](const char* proc_name, VkInstance instance, VkD
     CHECK_NONNULL(vk##F)
 
 VulkanInterface initVulkanInterface(bool protectedContent = false) {
-    const nsecs_t timeBefore = systemTime();
     VulkanInterface interface;
 
     VK_GET_PROC(EnumerateInstanceVersion);
@@ -604,9 +603,7 @@ VulkanInterface initVulkanInterface(bool protectedContent = false) {
     interface.isProtected = protectedContent;
     // funcs already initialized
 
-    const nsecs_t timeAfter = systemTime();
-    const float initTimeMs = static_cast<float>(timeAfter - timeBefore) / 1.0E6;
-    ALOGD("%s: Success init Vulkan interface in %f ms", __func__, initTimeMs);
+    ALOGD("%s: Success init Vulkan interface", __func__);
     return interface;
 }
 
@@ -662,25 +659,17 @@ static void sSetupVulkanInterface() {
     }
 }
 
-bool RenderEngine::canSupport(GraphicsApi graphicsApi) {
-    switch (graphicsApi) {
-        case GraphicsApi::GL:
-            return true;
-        case GraphicsApi::VK: {
-            if (!sVulkanInterface.initialized) {
-                sVulkanInterface = initVulkanInterface(false /* no protected content */);
-                ALOGD("%s: initialized == %s.", __func__,
-                      sVulkanInterface.initialized ? "true" : "false");
-            }
-            return sVulkanInterface.initialized;
-        }
-    }
-}
-
 namespace skia {
 
 using base::StringAppendF;
 
+bool SkiaVkRenderEngine::canSupportSkiaVkRenderEngine() {
+    VulkanInterface temp = initVulkanInterface(false /* no protected content */);
+    ALOGD("SkiaVkRenderEngine::canSupportSkiaVkRenderEngine(): initialized == %s.",
+          temp.initialized ? "true" : "false");
+    return temp.initialized;
+}
+
 std::unique_ptr<SkiaVkRenderEngine> SkiaVkRenderEngine::create(
         const RenderEngineCreationArgs& args) {
     std::unique_ptr<SkiaVkRenderEngine> engine(new SkiaVkRenderEngine(args));
diff --git a/libs/renderengine/skia/SkiaVkRenderEngine.h b/libs/renderengine/skia/SkiaVkRenderEngine.h
index 52bc500a5a..2e0cf45220 100644
--- a/libs/renderengine/skia/SkiaVkRenderEngine.h
+++ b/libs/renderengine/skia/SkiaVkRenderEngine.h
@@ -27,6 +27,8 @@ namespace skia {
 
 class SkiaVkRenderEngine : public SkiaRenderEngine {
 public:
+    // Returns false if Vulkan implementation can't support SkiaVkRenderEngine.
+    static bool canSupportSkiaVkRenderEngine();
     static std::unique_ptr<SkiaVkRenderEngine> create(const RenderEngineCreationArgs& args);
     ~SkiaVkRenderEngine() override;
 
diff --git a/libs/renderengine/tests/RenderEngineTest.cpp b/libs/renderengine/tests/RenderEngineTest.cpp
index 7b8eb8470f..4c18704f52 100644
--- a/libs/renderengine/tests/RenderEngineTest.cpp
+++ b/libs/renderengine/tests/RenderEngineTest.cpp
@@ -107,7 +107,7 @@ public:
 
     virtual std::string name() = 0;
     virtual renderengine::RenderEngine::GraphicsApi graphicsApi() = 0;
-    bool apiSupported() { return renderengine::RenderEngine::canSupport(graphicsApi()); }
+    virtual bool apiSupported() = 0;
     std::unique_ptr<renderengine::RenderEngine> createRenderEngine() {
         renderengine::RenderEngineCreationArgs reCreationArgs =
                 renderengine::RenderEngineCreationArgs::Builder()
@@ -131,6 +131,10 @@ public:
     renderengine::RenderEngine::GraphicsApi graphicsApi() override {
         return renderengine::RenderEngine::GraphicsApi::VK;
     }
+
+    bool apiSupported() override {
+        return skia::SkiaVkRenderEngine::canSupportSkiaVkRenderEngine();
+    }
 };
 
 class SkiaGLESRenderEngineFactory : public RenderEngineFactory {
@@ -140,6 +144,8 @@ public:
     renderengine::RenderEngine::GraphicsApi graphicsApi() {
         return renderengine::RenderEngine::GraphicsApi::GL;
     }
+
+    bool apiSupported() override { return true; }
 };
 
 class RenderEngineTest : public ::testing::TestWithParam<std::shared_ptr<RenderEngineFactory>> {
diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index 66af4ef8a3..517d78cb7c 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -819,10 +819,9 @@ void chooseRenderEngineType(renderengine::RenderEngineCreationArgs::Builder& bui
         builder.setThreaded(renderengine::RenderEngine::Threaded::YES)
                 .setGraphicsApi(renderengine::RenderEngine::GraphicsApi::VK);
     } else {
-        const auto kVulkan = renderengine::RenderEngine::GraphicsApi::VK;
-        const bool useVulkan = FlagManager::getInstance().vulkan_renderengine() &&
-                renderengine::RenderEngine::canSupport(kVulkan);
-        builder.setGraphicsApi(useVulkan ? kVulkan : renderengine::RenderEngine::GraphicsApi::GL);
+        builder.setGraphicsApi(FlagManager::getInstance().vulkan_renderengine()
+                                       ? renderengine::RenderEngine::GraphicsApi::VK
+                                       : renderengine::RenderEngine::GraphicsApi::GL);
     }
 }
 
-- 
2.48.1

